runtime/near-vm-runner/src/imports.rs

```diff
@@ -274,55 +274,150 @@ pub(crate) mod wasmer {
-        /// reference does not dangle. Still, would be great to fix this properly
-        /// one day.
-        logic: *mut (),
+    pub(crate) struct Wasmer2Imports<'vmlogic, 'vmlogic_refs> {
+        pub(crate) memory: VMMemory,
+        // Note: this same object is also referenced by the `metadata` field!
+        pub(crate) vmlogic: &'vmlogic mut VMLogic<'vmlogic_refs>,
+        pub(crate) metadata: Arc<ExportFunctionMetadata>,
+        pub(crate) protocol_version: ProtocolVersion,
     }
-    unsafe impl Send for NearWasmerEnv {}
-    unsafe impl Sync for NearWasmerEnv {}
-
-    pub(crate) fn build(
-        store: &wasmer::Store,
-        memory: wasmer::Memory,
-        logic: &mut VMLogic<'_>,
-        protocol_version: ProtocolVersion,
-    ) -> wasmer::ImportObject {
-        let env = NearWasmerEnv { logic: logic as *mut _ as *mut () };
-        let mut import_object = wasmer::ImportObject::new();
-        let mut namespace = wasmer::Exports::new();
-        namespace.insert("memory", memory);
 
-        macro_rules! add_import {
-            (
-              $func:ident < [ $( $arg_name:ident : $arg_type:ident ),* ] -> [ $( $returns:ident ),* ] >
-            ) => {
-                #[allow(unused_parens)]
-                fn $func(env: &NearWasmerEnv, $( $arg_name: $arg_type ),* ) -> Result<($( $returns ),*), VMLogicError> {
-                    const IS_GAS: bool = str_eq(stringify!($func), "gas");
-                    let _span = if IS_GAS {
-                        None
-                    } else {
-                        Some(tracing::trace_span!(target: "host-function", stringify!($func)).entered())
-                    };
-                    let logic: &mut VMLogic = unsafe { &mut *(env.logic as *mut VMLogic<'_>) };
-                    logic.$func( $( $arg_name, )* )
+    trait Wasmer2Type {
+        type Wasmer;
+        fn to_wasmer(self) -> Self::Wasmer;
+        fn ty() -> wasmer_types::Type;
+    }
+    macro_rules! wasmer_types {
+        ($($native:ty as $wasmer:ty => $type_expr:expr;)*) => {
+            $(impl Wasmer2Type for $native {
+                type Wasmer = $wasmer;
+                fn to_wasmer(self) -> $wasmer {
+                    self as _
                 }
+                fn ty() -> wasmer_types::Type {
+                    $type_expr
+                }
+            })*
+        }
+    }
+    wasmer_types! {
+        u32 as i32 => wasmer_types::Type::I32;
+        u64 as i64 => wasmer_types::Type::I64;
+    }
 
-                namespace.insert(stringify!($func), wasmer::Function::new_native_with_env(&store, env.clone(), $func));
-            };
+    macro_rules! return_ty {
+        () => {
+            type Ret = ();
+            fn make_ret() -> () {}
+        };
+        ($($return: ident),*) => {
+            #[repr(C)]
+            struct Ret($(<$return as Wasmer2Type>::Wasmer),*);
+            fn make_ret($($return: $return),*) -> Ret { Ret($($return.to_wasmer()),*) }
         }
```

Comment:
No strong opinion here, but I'd:

* Pass the `Ret` ident in, pretending that macros are hygenic
* use `$returns` rather than `$return` to be consistent with `add_import!` and to avoid messing mental syntax highlighting up: `$return` looks like some kind of hygenic `return`